keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Multinomial.java

```diff
@@ -123,26 +103,93 @@ private static int draw(KeanuRandom random, Double... categoryProbabilities) {
     }
 
     @Override
-    public DoubleTensor logProb(IntegerTensor k) {
-        long[] expectedShape = p.getShape();
-        TensorShapeValidation.checkAllShapesMatch(
-            String.format("Shape mismatch. k: %s, p: %s",
-                Arrays.toString(k.getShape()),
-                Arrays.toString(expectedShape)),
-            k.getShape(), expectedShape
-        );
-        Preconditions.checkArgument(
-            k.sum(0).elementwiseEquals(this.n).allTrue(),
-            String.format("Inputs %s must sum to n = %s", k, this.n)
+    public DoubleTensor logProb(IntegerTensor x) {
+        if (validationEnabled) {
+            validateBroadcastShapes(x.getShape(), n.getShape(), p.getShape());
+            validateX(x, n, p);
+        }
+
+        DoubleTensor gammaN = n.plus(1).toDouble().logGammaInPlace();
+        DoubleTensor xLogP = p.log().timesInPlace(x.toDouble()).sum(-1);
+        DoubleTensor gammaXs = x.plus(1).toDouble().logGammaInPlace().sum(-1);
+        return xLogP.plusInPlace(gammaN).minusInPlace(gammaXs);
+    }
+
+    private static void validateProbabilities(DoubleTensor p) {
+        if (p.isScalar()) {
+            throw new IllegalArgumentException("Probabilities must be a vector or a tensor with rank >= 1");
```

Comment:
throw a `io.improbable.keanu.tensor.TensorValueException` instead?

Here and elsewhere in the PR.

Reply:
I think IllegalArgumentException is more informative in this case. The TensorValueException doesn't really tell me more but instead really tells me less. This is definitely the case for checking hyper-parameter shapes or ranks. I thought about creating a ParameterValidationException but the value added is still pretty minimal.